src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentStepImpl.java

```diff
@@ -0,0 +1,234 @@
+package com.atlassian.bitbucket.jenkins.internal.deployments;
+
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironmentType;
+import com.atlassian.bitbucket.jenkins.internal.provider.JenkinsProvider;
+import com.google.common.annotations.VisibleForTesting;
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.model.Item;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.kohsuke.stapler.AncestorInPath;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.verb.POST;
+
+import javax.annotation.CheckForNull;
+import javax.inject.Inject;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.atlassian.bitbucket.jenkins.internal.deployments.DeploymentStepUtils.getOrGenerateEnvironmentKey;
+import static com.atlassian.bitbucket.jenkins.internal.deployments.DeploymentStepUtils.normalizeEnvironmentType;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.stripToNull;
+
+/**
+ * Step for configuring deployment notifications.
+ *
+ * @since deployments
+ */
+public class DeploymentStepImpl extends Step implements DeploymentStep {
+
+    private final String environmentName;
+    private final transient DeploymentStepExecutionProvider deploymentStepExecutionProvider;
+
+    private String environmentKey;
+    private BitbucketDeploymentEnvironmentType environmentType;
+    private String environmentUrl;
+
+    @DataBoundConstructor
+    public DeploymentStepImpl(@CheckForNull String environmentName) {
+        this.environmentName = stripToNull(environmentName);
+        this.deploymentStepExecutionProvider = getDescriptor().getDeploymentStepExecutionFactory();
+    }
+
+    @VisibleForTesting
```

Comment:
Same comment on this class. The descriptor is already public so we can mock it out on a spy rather than have a separate testing constructor.